Mixture model objects

bredelings opened the pull request:
This PR introduces RevBayes objects of type SiteMixtureModel.  A SiteMixtureModel is a collection of components, where each component has
* a weight, where the weights sum to 1
* root frequencies
* transition probabilities across each branch

This interface exposes only what the likelihood calculator needs to compute a likelihood.  This should improve modularity by hiding the other details .

The PR also makes it possible to construct substitution models as a single object.  This contrasts with the current method of constructing models like GTR+G4+INV, where we construct multiple objects and pass them as separate arguments to dnPhyloCTMC:
```
Q := fnGTR(er,pi) 
sr := fnDiscretizeGamma( alpha, alpha, 4, false )
p_inv ~ dnBeta(1,1)

seq ~ dnPhyloCTMC(tree=psi, Q=Q, siteRates=sr, pInv=p_inv, type="DNA")
```

With this PR, its possible to construct the model as a single object:
```
S := fnGTR(er,pi) |> fnGammaASRV(alpha,4,false) |> fnInvASRV(p_inv) 
seq ~ dnPhyloCTMC(tree=psi, Q=S, type="DNA")
```
The new functions have the following signatures:
* fnGammaASRV(MixtureModel<any> submodel, RealPos<any> alpha, Integer<any> n, Bool<any> median)
* fnInvASRV(MixtureModel<any> submodel, RealPos<any> pInv)
* fnMixtureASRV(SiteMixtureModel[]<any> models, Simplex<any> fractions, RealPos[]<any> rates)
* fnUnitMixture(RateGenerator<any> model, Simplex<any> rootFrequencies, RealPos<any> rate)
* fnScale(SiteMixtureModel<any> model, RealPos<any> rate) -> SiteMixtureModel
* fnScale(SiteMixtureModel[]<any> models, RealPos[]<any> rates) -> SiteMixtureModel[]

Its possible to make a  `SiteMixtureModel` out of any `RateGenerator` using the new `fnUnitMixture( )` function.  However, I've added a type conversion from `RateMatrix` to fnUnitMixture.  This uses the equilibrium frequences as the root frequencies automatically, so that its only necessary to use fnUnitMixture for non-equilibrium models.

Currently the model doesn't take advantage of the fact that RevBayes knows which sites are invariant.  This should be possible to add, and would result in a speedup.

In a second PR I propose to cut down the number of different model types that phyloCTMC has to handle by converting some of them to a SiteMixtureModel inside phyloCTMC. This has the benefit that old code will still work, because we will still accept e.g. siteRates and pInv.

hoehna commented on src/core/datatypes/trees/TopologyNode.h:
```diff
@@ -219,4 +219,6 @@ namespace RevBayesCore {
     };
 }
 
+std::pair<double,double> getStartEndAge(const RevBayesCore::TopologyNode& node);
```
Why is this a global method and not a member method?

bredelings replied:
As we talked about on the call, global methods are fine.

hoehna commented on src/core/datatypes/phylogenetics/substitution/MixtureModel.h:
```diff
@@ -0,0 +1,74 @@
+     * @since 2022-09-22, version 1.0
+     */
+
+    class MixtureModel: public Cloneable, public MemberObject< RbVector<RbVector<RbVector<double>>> >
```
Just my personal view:
Could you rename these into SubstitutionMixtureModel? I think that is more appropriate. Note that we have other types of mixture models too, e.g., https://github.com/revbayes/revbayes/blob/7b872a7095c5342297ad4a1f254b29d81cc68040/src/core/distributions/AnalyticalMixtureDistribution.h#L26

bredelings replied:
Yeah, good point.

hoehna commented on src/core/distributions/phylogenetics/substitution/AbstractPhyloCTMCSiteHomogeneous.h:
```diff
@@ -1535,7 +1503,7 @@ bool RevBayesCore::AbstractPhyloCTMCSiteHomogeneous<charType>::recursivelyDrawSt
 
     // get the rate matrix for this branch (or site if using a mixture of matrices over sites)
-    RateMatrix_JC jc(this->num_chars);
+    RateMatrix_JC jc(this->num_chars); // BAD!!
```
Yes, any better ideas?

bredelings replied:
I just want to mark this to work on later.

hoehna commented on src/revlanguage/functions/phylogenetics/ratematrix/Func_GammaRateModel.cpp:
```diff
@@ -0,0 +1,186 @@
+std::string Func_GammaRateModel::getFunctionName( void ) const
+{
+    // create a name variable that is the same for all instance of this class
+    std::string f_name = "fnGamma";
```
This definitely needs a different names. Better would be `fnGammaRates` or `fnGammaASRV`. For a function with name `fnGamme` I would expect either the Gamma function, or the Gamma distribution (e.g., pdf of Gamma).

bredelings replied:
Based on the poll, I've changed `fnGamma` -> `fnGammaASRV`.

hoehna commented on src/revlanguage/functions/phylogenetics/ratematrix/Func_InvModel.cpp:
```diff
@@ -0,0 +1,153 @@
+std::string Func_InvModel::getFunctionName( void ) const
+{
+    // create a name variable that is the same for all instance of this class
+    std::string f_name = "fnInv";
```
Same as for `fnGamma`, I would suggest a more descriptive name. Perhaps `fnInvSites`?

hoehna reviewed (commented):
This looks in general very good. My only comment is that perhaps not everyone would have the same association when you talk about a mixture model. Therefore, not everyone would associate this with a substitution mixture model. Does this automatically also work for a, say, branch rate mixture model? I simply suggest to make the names more clear, if possible the C++ class names, but at the minimum the Rev function names (these will lead to confusion otherwise).

bredelings commented:
Things to handle in round 1:
- check no speed slowdown
- handle other phyloCTMC variants
- for handling fnEpoch, we could force people to 
- handle speedup for fnINV.

bjoelle commented on tests/test_mixtures/scripts/mcmc_mixture2.rev:
is it possible to add a test that the model defined in the "old" way and the mixture model give the same likelihood ? just to avoid any drift later

bredelings replied:
We should.  We could add testiphy tests perhaps, since that is designed to test likelihoods.

So far the "old" and "new" scripts actually yield the same MCMC trace, so they are giving the same likelihoods.  But that is not a general test framework....

bredelings replied:
I changed the stepping stone test to use the "new" way.  It doesn't change the output except that it doesn't print out the gamma rates anymore because we just compute fnGammaASRV directly.

I also added a test in test_likelihood that the "old" way and the "new" way yield the same likelihood to 14 digits.

bjoelle reviewed (commented):
overall this seems good

it's missing quite a bit of documentation in the new classes, in particular I'm having a really hard time parsing out the hierarchy of the mixture models
it _seems_ that for the complete package you get a ConcreteMixtureModel that points to a BranchRatesModel that points to a Concrete that points to (maybe ?) a Concrete with the actual transition matrix and a Unit with the pInv ?

also I think this would need more tests in particular for validating that a different order of chaining the functions gives the same result (if that is the intention ? otherwise the proper order need to be enforced somewhere I think)

bredelings commented on src/revlanguage/functions/phylogenetics/ratematrix/Func_ConvertVectorRateMatrix.cpp:
```diff
@@ -0,0 +1,133 @@
+std::string Func_ConvertVectorRateMatrix::getFunctionName( void ) const
+{
+    // create a name variable that is the same for all instance of this class
+    std::string f_name = "_RateGenerator[]2SiteMixtureModel[]";
```
Revbayes searches names of the form "_TypeA2TypeB" when it looks for conversion functions.  So the "2" is a magic symbol.

bredelings commented on src/core/datatypes/phylogenetics/ratematrix/ConcreteTimeReversibleRateMatrix.cpp:
```diff
@@ -132,8 +132,9 @@ void ConcreteTimeReversibleRateMatrix::update( void )
 
-        // rescale
-        rescaleToAverageRate( 1.0 );
+        // rescale - Problem!
```
The problem is fixed by only rescaling to a specific rate when the optional is non-empty.  I will fix the comment.

bredelings replied:
Fixed.

bredelings commented on src/core/datatypes/RbVectorImpl.h:
```diff
@@ -267,6 +268,7 @@ namespace RevBayesCore {
         RbIterator<valueType>                               find(const valueType &x) { return RbIterator<valueType>( std::find(this->values.begin(), this->values.end(), &x) ); }
         RbConstIterator<valueType>                          find(const valueType &x) const { return RbConstIterator<valueType>( std::find(this->values.begin(), this->values.end(), &x) ); }
         size_t                                              size(void) const { return this->values.size(); }
+        void                                                setElement(int i, const valueType& v) { delete values[i]; values[i] = Cloner<valueType, IsDerivedFrom<valueType, Cloneable>::Is >::createClone(v);}
```
Not a dumb question!  RbVectorImpl has *two* implementations.  The first implementation derives from `std::vector<T>`, but does not work if `T` is an abstract class.  So this implementation gets used for e.g. `RbVector<double>`.

  The second implementation uses `std::vector<T*>` but the interface make it look like a collection of `T`s.  This works for abstract classes.

bredelings commented on src/revlanguage/functions/phylogenetics/ratematrix/Func_ConvertRateMatrix.cpp:
```diff
@@ -0,0 +1,124 @@
+    if (auto matrix = dynamic_cast<const Core::RateMatrix*>(&model))
+        return new Core::UnitMixtureModel(*matrix, matrix->getStationaryFrequencies());
+    else
+        throw RbException()<<"_RateMatrix2MixtureModel: model is not a RateMatrix!";
```
The problem is that RateGenerator does not necessary have stationary frequencies.  Only RateMatrix does.

If you write `x = [ fnJC(4) ]` then apparently `x` has type `RateGenerator[]` and not `RateMatrix[]`.  Maybe this could be fixed, but I have not looked into it yet...

bredelings replied:
Oh, actually the arguments DO enforce it.  So, the exception should not occur.  I can put a note in.
But in `Func_ConvertVectoRateMatrix` I am not sure how to enforce this .

bredelings replied:
It does -- this is just to catch a "this shouldn't happen" case.

bredelings commented on src/revlanguage/functions/phylogenetics/ratematrix/Func_GammaRateModel.cpp:
```diff
@@ -0,0 +1,186 @@
+// well, ratemodel should perhaps be its OWN interface!
+
+// how about things like the m3 model?
+// also, could we create this using a FUNCTION?
```
Yeah, I can remove.

bredelings replied:
Fixed.

bredelings commented on src/revlanguage/functions/phylogenetics/ratematrix/Func_GammaRateModel.cpp:
```diff
@@ -0,0 +1,186 @@
+{
+    using namespace RevBayesCore;
+
+    double b = a;
```
I can add a comment.
Gamma is normally Gamma(a,b).   But in this case we set b = a so that the mean will be 1.

bredelings commented on src/revlanguage/functions/phylogenetics/ratematrix/Func_InvModel.cpp:
```diff
@@ -0,0 +1,153 @@
+// well, ratemodel should perhaps be its OWN interface!
+
+// how about things like the m3 model?
+// also, could we create this using a FUNCTION?
```
Will fix

bredelings replied:
Fixed.

bjoelle commented:
ok, so my impression was that most new functions would do some sort of nesting, so you would end up with a final model which would actually be a series of nested models downwards (even in a simple-ish case like one relaxed clock + one GTR+G)
so I think I mostly got what each individual class does (although it's still worth documenting :) but I was less sure about what the end result would be, if that makes sense ?
that could mostly be sorted by documentation of inputs/outputs in the functions I think, although ultimately it might be worth doing a small tutorial on the user side too

bredelings commented on src/revlanguage/functions/phylogenetics/ratematrix/Func_GammaRateModel.cpp:
```diff
@@ -0,0 +1,186 @@
+
+    if ( !rules_set )
+    {
+        argumentRules.push_back( new ArgumentRule( "submodel"       , SiteMixtureModel::getClassTypeSpec(), "Sub-model.", ArgumentRule::BY_CONSTANT_REFERENCE, ArgumentRule::ANY ) );
```
So, it actually really modeled on the same interface.  When people use Gamma-distributed rate variation across sites, the mean of the distribution is constrained to 1, and there is only 1 free parameter, which is traditionally called alpha.  (If you don't constraint the mean to 1, then you end up redefining the meaning of branch lengths by a constant factor.)

bredelings commented:
I will add documentation of the function at the Rev language level.  That will show how the nesting is supposed to work.

In some cases multiple nesting orders work, but one is preferred.

bredelings commented:
I can look at writing a tutorial after the PR is merged.  I probably also need to do one for Codon models too.

bredelings commented on src/revlanguage/functions/phylogenetics/ratematrix/Func_InvModel.cpp:
```diff
@@ -0,0 +1,153 @@
+    sub_models.push_back( unique_ptr<RevBayesCore::SiteMixtureModel>(new RevBayesCore::UnitMixtureModel(INV)) );
+    
+    auto inv_model = new RevBayesCore::ConcreteMixtureModel(sub_models, fractions);
+    inv_model->setRate(1);
```
I've changed this so that instead of setting the rate to 1, we set the rate equal to the submodel rate.  This is equivalent to what revbayes is doing now.

bredelings commented on src/revlanguage/functions/phylogenetics/ratematrix/Func_GammaRateModel.cpp:
```diff
@@ -0,0 +1,186 @@
+    {
+        argumentRules.push_back( new ArgumentRule( "submodel"       , SiteMixtureModel::getClassTypeSpec(), "Sub-model.", ArgumentRule::BY_CONSTANT_REFERENCE, ArgumentRule::ANY ) );
+        argumentRules.push_back( new ArgumentRule( "alpha"          , RealPos::getClassTypeSpec(), "The alpha parameter of the gamma distribution.", ArgumentRule::BY_CONSTANT_REFERENCE, ArgumentRule::ANY ) );
+        argumentRules.push_back( new ArgumentRule( "n"              , Integer::getClassTypeSpec(), "The number of bins to approximate the gamma distirbution.", ArgumentRule::BY_CONSTANT_REFERENCE, ArgumentRule::ANY, new Integer(4) ) );
```
Fixed.

bredelings commented on src/revlanguage/distributions/phylogenetics/character/Dist_PhyloCTMCClado.cpp:
```diff
@@ -349,6 +365,11 @@ RevBayesCore::TypedDistribution< RevBayesCore::AbstractHomologousDiscreteCharact
             
             dist->setRateMatrix( rm );
         }
+        else if ( q->getRevObject().isType( SiteMixtureModel::getClassTypeSpec() ) )
```
Unduplicated the shared code in dnPhyloCTMC and dnPhyloCTMCClado by defining a template member function for each class.  The problem was that we wanted to create a different distribution type depending on the value of a string, and then do the same things on each distribution.

bredelings commented on src/core/distributions/MixtureDistribution.h:
```diff
@@ -277,13 +286,13 @@ void RevBayesCore::MixtureDistribution<mixtureType>::touchSpecialization( const
-        
-        if ( this->dag_node != NULL )
-        {
-            this->dag_node->touchAffected();
```
:grimacing: Fixed.

bredelings commented:
@bjoelle I'm starting to add documentation -- see the help for fnGammaASRV.  Does this clarify some things?

bredelings commented:
I've eliminate fnBranchRates( ) in favor of just setting 'branchRates = branch_rates' in dnPhyloCTMC( ).  The problem was that `fnBranchRates(branch_rates)` was substantially slower when MCMC moves change just one or two rates at a time.  I don't see a way to tell dnPhyloCTMC that the transition matrix changed only on one branch using a SiteMixtureModel.

bredelings commented:
@bjoelle OK, I think this is ready.

bjoelle commented on src/revlanguage/functions/phylogenetics/ratematrix/Func_InvModel.h:
```diff
@@ -0,0 +1,52 @@
+class TypeSpec;
+
+    /**
+     * The RevLanguage wrapper of the Gamma model of rates-across-sites.
```
that is not correct, is it ?

bredelings replied:
fixed

bjoelle commented on src/revlanguage/functions/phylogenetics/ratematrix/Func_ConvertRateMatrix.h:
```diff
@@ -0,0 +1,52 @@
+class TypeSpec;
+
+    /**
+     * The RevLanguage wrapper of the Gamma model of rates-across-sites.
```
that is not correct, is it ?

bjoelle commented on src/revlanguage/functions/phylogenetics/ratematrix/Func_UnitMixture.h:
```diff
@@ -0,0 +1,52 @@
+class TypeSpec;
+
+    /**
+     * The RevLanguage wrapper of the Gamma model of rates-across-sites.
```
that is not correct, is it ?

bredelings replied:
fixed

bjoelle reviewed (changes requested):
thank you for the changes ! I've added a few comments on typos in the help & class documentation, but I have no other changes to add

bredelings commented on src/core/help/RbHelpDatabase.cpp:
```diff
@@ -1359,6 +1360,43 @@ Q := fnGTR(er,pi))");
+return a model with m*n components.
+
+The continuous Gamma distribution is approximated with a mixture distribution
+over n discrete rates, each with probability 1/4.  The Gamma distribution is
```
fixed

bredelings commented on src/core/help/RbHelpDatabase.cpp:
```diff
@@ -1359,6 +1360,43 @@ Q := fnGTR(er,pi))");
+        - If alpha = 1, then the rate is exponentially distributed.  Rate variation is substantial.
+        - As alpha approaches zero, many sites have rate 0, and many sites have a high rate.
+
+RateMatrix and RateGenerator submodels will automatically converted to a
```
fixed

bredelings commented on src/core/help/RbHelpDatabase.cpp:
```diff
@@ -1371,12 +1409,59 @@ Q := fnHKY(kappa,pi))");
+	help_strings[string("fnInvASRV")][string("description")] = string(R"(Add an invariable-sites component to a submodel.)");
+	help_strings[string("fnInvASRV")][string("details")] = string(R"(This model specifies that some fraction pInv of sites are invariable.
+If the submodel has m components, this function will return a model with
+n+1 components.)");
```
fixed

wrightaprilm commented on help/md/fnUnitMixture.md:
```diff
@@ -0,0 +1,33 @@
+## see_also
+## example
+        M := fnUnitMixture( fnJC(4) )
+        M := fnJC(4) |> fnUnitMixture()  # nested functions can be expressed using pipes.
```
I find these examples a little unclear the documentation from the RB help function implies you can specify multiple models, but this is not shown as an example.

bredelings replied:
OK, I change this to use fnJC, fnGTR, and fnDECRateMatrix

wrightaprilm reviewed (changes requested):
I did some checking, and made scripts that should be equivalent for among-character rate variation between the 'standard' way we've been implementing this, and the mixture models. I've verified that the posteriors returned via the two methods are the same. Very technically sound.

I've made some suggestions on language in the help files. Well on the way! Good work, Ben!

bredelings commented on help/md/fnGammaASRV.md:
```diff
@@ -0,0 +1,50 @@
+Add Gamma-distributed across-site rate variation (ASRV) to a submodel.
+
+## details
+Each site evolves according to the specified submodel, but at an unknown rate
```
I guess so.  Sure.

FWIW, you can think of the Gamma site model as a model that takes a real number (alpha). an integer (n), and a model (the site model) as parameters.  When a model is parameterized by another model, I've been calling that a sub-model.  Do you any suggestions for something else to call it?

I do realize that this is kind of a conceptual jump (models parameterized by other models) so I think its fine to just say "site model" here.

wrightaprilm replied:
It's not a super strong feeling - because I agree submodel makes sense in the context you've described. But it also gets used in hierarchical modeling, where you might have different processes describing different data sources. So if you want to leave it, I think it's fine, but for the precision of the example, site model might be better.

bredelings replied:
OK, staying with "site model" and "site model parameter".

bredelings commented on help/md/fnGammaASRV.md:
```diff
@@ -0,0 +1,50 @@
+
+## details
+Each site evolves according to the specified submodel, but at an unknown rate
+that is Gamma distributed. If the submodel has m components, this function will
```
This looks unfinished...

wrightaprilm replied:
It's fine - you solved it with another edit.

bredelings commented on help/md/fnGammaASRV.md:
```diff
@@ -0,0 +1,50 @@
+## example
+        alpha ~ dnExp(1/10) # Don't put too much prior belief on high rate variation.
+        M := fnGammaASRV( fnGTR(er, pi), alpha, 4)
+        M := fnGTR(er,pi) |> fnGammaASRV(alpha, 4)  # Nested functions can be expressed using pipes.
```
I tried to reformulate the help to follow these suggestions.

bredelings commented on help/md/fnGammaASRV.md:
```diff
@@ -0,0 +1,50 @@
+        weights ~ dnDirichlet([1,1])
+        M := fnMixtureASRV([fnF81(pi1),fnF81(pi2)],weights) |> fnGammaASRV(alpha) |> fnInvASRV(p_inv)
+
+        # Rate is a product of two Gamma-distributed variables.  For illustration only -- NOT recommended.
```
I took out the GTR+Gamma+Gamma example.

wrightaprilm commented on help/md/fnMixtureASRV.md:
```diff
@@ -0,0 +1,38 @@
+## name
+fnMixtureASRV
+## title
+fnMixtureASRV
```
I like the new changes. They read well.

bredelings commented on help/md/fnGammaASRV.md:
```diff
@@ -0,0 +1,50 @@
+fnMixtureASRV
+fnDiscretizeGamma
+## example
+        alpha ~ dnExp(1/10) # Don't put too much prior belief on high rate variation.
```
Changed.

bredelings commented:
I had a brainstorm and refactored the SiteMixtureModel class to actually consist of a vector of SiteModel and a vector<double> for the weights.  This allowed eliminating the ConcreteSiteModel class since SiteMixtureModel is no longer abstract.

I also removed the rates parameter from fnMixtureASRV(models,weights,rates).  So no its just fnMixtureASRV(models,weights), as it should be.

I created a new fnScale(models,rates) to scale model[i] by rate[i].  As a bonus, if you supply a single model but a vector of rates, it automatically replicates the single model to a vector of the same.  So now you can write `fnScale(Q, rates)` instead of `fnScale([Q,Q,Q,Q],rates)`, although both work.

Should I rename fnScale to fnScaleASRV?  I am unsure.

wrightaprilm commented:
> Should I rename fnScale to fnScaleASRV?

I guess my question would be, how do you want people to use this? If the ASRV rates are always mandatory, I would rename. If the ASRV rates are optional, I would not.
